File: NodeJs_TypeScript_JavaScript_Server/Project_NodeJs_VSCode/src/base/crypto/util/CryptoUtil.ts
import * as random_seed from "random-seed";
import { LogicException } from "base/exception/LogicException";
import { Salsa20Algorithm } from "../algorithm/Salsa20Algorithm";
import { EncryptCipher } from "../cipher/EncryptCipher";
import { IEncryptCipher } from "../cipher/IEncryptCipher";
import { DecryptCipher } from "../cipher/DecryptCipher";
import { IDecryptCipher } from "../cipher/IDecryptCipher";

export class CryptoUtil {

    public static Algorithms_Salsa20: string = "_salsa20"

    public static RNG = random_seed.create((new Date()).getTime().toString(8));
    
    public static MD5_LENGTH: number = 32;
    public static HEXADECIMAL_CHARACTERS: string = "0123456789ABCDEF";
  
    public static randomHexString(len: number): string {
        let s: string = '';
        for(let i: number=0;i<len;i++) {
            s += CryptoUtil.HEXADECIMAL_CHARACTERS.charAt(CryptoUtil.RNG.range(CryptoUtil.HEXADECIMAL_CHARACTERS.length));
        }
        
        return s;
    }
  
    public static newEncryptCipher(algorithmType: string, key: Uint8Array, iv: Uint8Array): IEncryptCipher {
        if(algorithmType == CryptoUtil.Algorithms_Salsa20) {
            if(key.byteLength != Salsa20Algorithm.KEY_SIZE || iv.byteLength != Salsa20Algorithm.IV_SIZE) {
                throw new LogicException("Err.");
            }

            let algorithm: Salsa20Algorithm = new Salsa20Algorithm(key, iv);
    
            let cipher: EncryptCipher = new EncryptCipher(algorithm);
    
            return cipher;
        }
    
        throw new LogicException("Err.");
    }

    public static newDecryptCipher(algorithmType: string, key: Uint8Array, iv: Uint8Array): IDecryptCipher {
        if(algorithmType == CryptoUtil.Algorithms_Salsa20) {
            if(key.byteLength != Salsa20Algorithm.KEY_SIZE || iv.byteLength != Salsa20Algorithm.IV_SIZE) {
                throw new LogicException("Err.");
            }
    
            let algorithm: Salsa20Algorithm = new Salsa20Algorithm(key, iv);
    
            let cipher: DecryptCipher = new DecryptCipher(algorithm);
    
            return cipher;
        }
    
        throw new LogicException("Err.");
    }

}

File: CPP_Joc_Windows_Android/SH_Client_Win_Cpp_Cmake/App/src/base/crypto/util/CryptoUtil.cpp
#include <base/crypto/algorithm/Salsa20Algorithm.h>
#include "CryptoUtil.h"
#include <base/exceptions/LogicException.h>
#include <base/crypto/cipher/EncryptCipher.h>
#include <base/crypto/cipher/DecryptCipher.h>

using namespace base::crypto;

const std::string CryptoUtil::Algorithms::Salsa20 = "_salsa20";

std::shared_ptr<base::crypto::IEncryptCipher> CryptoUtil::newEncryptCipher(
    std::string algorithmType,
    const char* key, int keySize, const char* iv, int ivSize)
{
    if(algorithmType == CryptoUtil::Algorithms::Salsa20) {
        if(keySize != Salsa20Algorithm::KEY_SIZE || ivSize != Salsa20Algorithm::IV_SIZE) {
            throw LogicException(LOC);
        }

        std::shared_ptr<base::crypto::Salsa20Algorithm> algorithm = std::make_shared<base::crypto::Salsa20Algorithm>(reinterpret_cast<const uint8_t*>(key), reinterpret_cast<const uint8_t*>(iv));

        std::shared_ptr<base::crypto::EncryptCipher> cipher = std::make_shared<base::crypto::EncryptCipher>(algorithm);

        return cipher;
    }

    throw LogicException(LOC);
}

std::shared_ptr<base::crypto::IDecryptCipher> CryptoUtil::newDecryptCipher(
    std::string algorithmType,
    const char* key, int keySize, const char* iv, int ivSize)
{
    if(algorithmType == CryptoUtil::Algorithms::Salsa20) {
        if(keySize != Salsa20Algorithm::KEY_SIZE || ivSize != Salsa20Algorithm::IV_SIZE) {
            throw LogicException(LOC);
        }

        std::shared_ptr<base::crypto::Salsa20Algorithm> algorithm = std::make_shared<base::crypto::Salsa20Algorithm>(reinterpret_cast<const uint8_t*>(key), reinterpret_cast<const uint8_t*>(iv));

        std::shared_ptr<base::crypto::DecryptCipher> cipher = std::make_shared<base::crypto::DecryptCipher>(algorithm);

        return cipher;
    }

    throw LogicException(LOC);
}
